### PR TITLE
feat(perception_evaluator): counts objects within detection range 

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

- feat(perception_evaluator): counts objects within detection range 
  - move from https://github.com/autowarefoundation/autoware.universe/pull/6761 
- refactoring
- fix (e.g. dynamic reconfigure)

image

image

## Tests performed

psim

https://github.com/autowarefoundation/autoware.universe/assets/39142679/02082d78-b973-4b1d-b837-3fc115c5b4e0

unit test (with new tests)
```
1: [==========] Running 38 tests from 1 test suite.
1: [----------] Global test environment set-up.
1: [----------] 38 tests from EvalTest
1: [ RUN      ] EvalTest.testLateralDeviation_deviation0
1: [       OK ] EvalTest.testLateralDeviation_deviation0 (1449 ms)
1: [ RUN      ] EvalTest.testLateralDeviation_deviation1
1: [       OK ] EvalTest.testLateralDeviation_deviation1 (1452 ms)
1: [ RUN      ] EvalTest.testLateralDeviation_oscillation
1: [       OK ] EvalTest.testLateralDeviation_oscillation (2468 ms)
1: [ RUN      ] EvalTest.testLateralDeviation_distortion
1: [       OK ] EvalTest.testLateralDeviation_distortion (2446 ms)
1: [ RUN      ] EvalTest.testLateralDeviation_deviation0_PEDESTRIAN
1: [       OK ] EvalTest.testLateralDeviation_deviation0_PEDESTRIAN (1431 ms)
1: [ RUN      ] EvalTest.testYawDeviation_deviation0
1: [       OK ] EvalTest.testYawDeviation_deviation0 (1425 ms)
1: [ RUN      ] EvalTest.testYawDeviation_deviation1
1: [       OK ] EvalTest.testYawDeviation_deviation1 (1443 ms)
1: [ RUN      ] EvalTest.testYawDeviation_oscillation
1: [       OK ] EvalTest.testYawDeviation_oscillation (2468 ms)
1: [ RUN      ] EvalTest.testYawDeviation_distortion
1: [       OK ] EvalTest.testYawDeviation_distortion (2449 ms)
1: [ RUN      ] EvalTest.testYawDeviation_oscillation_rotate
1: [       OK ] EvalTest.testYawDeviation_oscillation_rotate (2433 ms)
1: [ RUN      ] EvalTest.testYawDeviation_distortion_rotate
1: [       OK ] EvalTest.testYawDeviation_distortion_rotate (2458 ms)
1: [ RUN      ] EvalTest.testYawDeviation_deviation0_PEDESTRIAN
1: [       OK ] EvalTest.testYawDeviation_deviation0_PEDESTRIAN (1442 ms)
1: [ RUN      ] EvalTest.testPredictedPathDeviation_deviation0
1: [       OK ] EvalTest.testPredictedPathDeviation_deviation0 (1427 ms)
1: [ RUN      ] EvalTest.testPredictedPathDeviation_deviation1
1: [       OK ] EvalTest.testPredictedPathDeviation_deviation1 (2445 ms)
1: [ RUN      ] EvalTest.testPredictedPathDeviation_deviation2
1: [       OK ] EvalTest.testPredictedPathDeviation_deviation2 (1456 ms)
1: [ RUN      ] EvalTest.testPredictedPathDeviation_deviation0_PEDESTRIAN
1: [       OK ] EvalTest.testPredictedPathDeviation_deviation0_PEDESTRIAN (1454 ms)
1: [ RUN      ] EvalTest.testPredictedPathDeviationVariance_deviation0
1: [       OK ] EvalTest.testPredictedPathDeviationVariance_deviation0 (12483 ms)
1: [ RUN      ] EvalTest.testPredictedPathDeviationVariance_deviation1
1: [       OK ] EvalTest.testPredictedPathDeviationVariance_deviation1 (12468 ms)
1: [ RUN      ] EvalTest.testPredictedPathDeviationVariance_deviationIncreasing
1: [       OK ] EvalTest.testPredictedPathDeviationVariance_deviationIncreasing (12490 ms)
1: [ RUN      ] EvalTest.testPredictedPathDeviationVariance_deviationOscillating
1: [       OK ] EvalTest.testPredictedPathDeviationVariance_deviationOscillating (12467 ms)
1: [ RUN      ] EvalTest.testYawRate_0
1: [       OK ] EvalTest.testYawRate_0 (1428 ms)
1: [ RUN      ] EvalTest.testYawRate_01
1: [       OK ] EvalTest.testYawRate_01 (3042 ms)
1: [ RUN      ] EvalTest.testYawRate_minus_01
1: [       OK ] EvalTest.testYawRate_minus_01 (3041 ms)
1: [ RUN      ] EvalTest.testYawRate_1
1: [       OK ] EvalTest.testYawRate_1 (3051 ms)
1: [ RUN      ] EvalTest.testYawRate_minus_1
1: [       OK ] EvalTest.testYawRate_minus_1 (3056 ms)
1: [ RUN      ] EvalTest.testYawRate_5
1: [       OK ] EvalTest.testYawRate_5 (3050 ms)
1: [ RUN      ] EvalTest.testYawRate_minus_5
1: [       OK ] EvalTest.testYawRate_minus_5 (3035 ms)
1: [ RUN      ] EvalTest.testAverageObjectsCount_CAR
1: [       OK ] EvalTest.testAverageObjectsCount_CAR (1428 ms)
1: [ RUN      ] EvalTest.testAverageObjectsCount_UNKNOWN
1: [       OK ] EvalTest.testAverageObjectsCount_UNKNOWN (1445 ms)
1: [ RUN      ] EvalTest.testAverageObjectsCount_MultiClass
1: [       OK ] EvalTest.testAverageObjectsCount_MultiClass (1444 ms)
1: [ RUN      ] EvalTest.testIntervalAverageObjectsCount_CAR
1: [       OK ] EvalTest.testIntervalAverageObjectsCount_CAR (1426 ms)
1: [ RUN      ] EvalTest.testIntervalAverageObjectsCount_UNKNOWN
1: [       OK ] EvalTest.testIntervalAverageObjectsCount_UNKNOWN (1426 ms)
1: [ RUN      ] EvalTest.testIntervalAverageObjectsCount_MultiClass
1: [       OK ] EvalTest.testIntervalAverageObjectsCount_MultiClass (1430 ms)
1: [ RUN      ] EvalTest.testTotalObjectsCount_CAR
1: [       OK ] EvalTest.testTotalObjectsCount_CAR (1447 ms)
1: [ RUN      ] EvalTest.testTotalObjectsCount_UNKNOWN
1: [       OK ] EvalTest.testTotalObjectsCount_UNKNOWN (1453 ms)
1: [ RUN      ] EvalTest.testTotalObjectsCount_DifferentCAR
1: [       OK ] EvalTest.testTotalObjectsCount_DifferentCAR (1440 ms)
1: [ RUN      ] EvalTest.testTotalObjectsCount_OutsideCAR
1: [       OK ] EvalTest.testTotalObjectsCount_OutsideCAR (1439 ms)
1: [ RUN      ] EvalTest.testTotalObjectsCount_OutsideHeightCAR
1: [       OK ] EvalTest.testTotalObjectsCount_OutsideHeightCAR (1441 ms)
1: [----------] 38 tests from EvalTest (115577 ms total)
1:
1: [----------] Global test environment tear-down
1: [==========] 38 tests from 1 test suite ran. (115577 ms total)
1: [  PASSED  ] 38 tests.
```

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: evaluator/perception_online_evaluator/README.md [hunk 1]
```diff
@@ -99,14 +163,24 @@ Calculates the yaw rate of an object based on the change in yaw angle from the p
...
+| `selected_metrics`                                     | List         | Metrics to be evaluated, such as lateral deviation, yaw deviation, and predicted path deviation.                                                |
+| `smoothing_window_size`                                | Integer      | Determines the window size for smoothing path, should be an odd number.                                                                         |
+| `prediction_time_horizons`                             | list[double] | Time horizons for prediction evaluation in seconds.                                                                                             |
+| `stopped_velocity_threshold`                           | double       | threshold velocity to check if vehicle is stopped                                                                                               |
+| `detection_radius_list`                                | double       | Detection radius for objects to be evaluated.(used for objects count only)                                                                      |
```
Comment (kyoichi-sugahara) [inline evaluator/perception_online_evaluator/README.md:172]: list[double] ?

File: evaluator/perception_online_evaluator/src/perception_online_evaluator_node.cpp [hunk 2]
```diff
@@ -292,6 +273,7 @@ rcl_interfaces::msg::SetParametersResult PerceptionOnlineEvaluatorNode::onParame
   result.successful = true;
   result.reason = "success";
 
+  std::cerr << "Parameters updated" << std::endl;
```
Comment (kyoichi-sugahara) [inline evaluator/perception_online_evaluator/src/perception_online_evaluator_node.cpp:276]: left debug print?

File: evaluator/perception_online_evaluator/README.md (hunk 1 above)
Comment (kosuke55) [reply evaluator/perception_online_evaluator/README.md]: right, thnaks! d29625fa711d05dbfc0ad183bff4d52c76483296

File: evaluator/perception_online_evaluator/src/perception_online_evaluator_node.cpp (hunk 2 above)
Comment (kosuke55) [reply evaluator/perception_online_evaluator/src/perception_online_evaluator_node.cpp]: d29625fa711d05dbfc0ad183bff4d52c76483296

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM!
My brain can't fully catch up 2D map too dificult :cry: 
I appreciate how clean the process flow has become! I will close my PR.

File: evaluator/perception_online_evaluator/src/metrics/detection_count.cpp [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+      const bool is_within_detection_radius = distance_to_base_link < range.radius;
+      const bool is_below_detection_height = pose_out.pose.position.z < range.height;
+      if (is_within_detection_radius && is_below_detection_height) {
+        updateDetectionMap(uuid, label, range.toString(), timestamp);
+      }
+    }
+  }
+}
+
+void DetectionCounter::initializeDetectionMap()
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply evaluator/perception_online_evaluator/src/metrics/detection_count.cpp:101]: this initialization is for publishing the count values as 0.

[the remark this answers is not in the document]
Comment (kosuke55) [reply evaluator/perception_online_evaluator/src/metrics/detection_count.cpp:101]: for exmaple, 
If there are no initialization, this for loop is skipped in `getAverageCount`

```
  for (auto & [classification, range_map] : time_series_counts_) {
    if (
      parameters_->object_parameters.find(classification) == parameters_->object_parameters.end() ||
      !parameters_->object_parameters.at(classification).check_interval_average_objects_count) {
      continue;
    }
    for (auto & [range, timestamps] : range_map) {
```

[the remark this answers is not in the document]
Comment (kosuke55) [reply evaluator/perception_online_evaluator/src/metrics/detection_count.cpp:101]: what do you think? you think publish 0 count value is neccessary?

[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply evaluator/perception_online_evaluator/src/metrics/detection_count.cpp:101]: @kosuke55 
> this initialization is for publishing the count values as 0.

Ah, I fully understand!!! thanks.
In this case, initialization is necessary!

> what do you think? you think publish 0 count value is neccessary?

Yes if the class is not detected, 0 value is better to be published.

**Pull request merged**